src/main/kotlin/io/openfuture/snapshot/snapshot/TransferEventBasedSnapshotCreator.kt

```diff
@@ -0,0 +1,63 @@
+package io.openfuture.snapshot.snapshot
+
+import io.openfuture.snapshot.domain.WalletState
+import org.web3j.protocol.core.methods.response.EthLog
+import org.web3j.utils.Convert
+import java.math.BigDecimal
+import java.math.BigInteger
+
+class TransferEventBasedSnapshotCreator(nodeAddress: String) : CommonSnapshotCreator(nodeAddress) {
+
+    override fun snapshot(contractAddress: String, fromBlock: Int, toBlock: Int): Set<WalletState> {
+        val walletStateMap: MutableMap<String, BigDecimal> = mutableMapOf()
+
+        for (blockNumber in fromBlock..toBlock step BATCH_SIZE) {
+            var nextBatch = blockNumber + BATCH_SIZE - 1
+            if (nextBatch >= toBlock) nextBatch = toBlock
+
+            println("Batch fetching addresses from $blockNumber to $nextBatch blocks")
+            val logs = getTransferLogs(contractAddress, blockNumber, nextBatch)
+            val transfers = fetchTransferEvents(logs)
+            calculateBalances(walletStateMap, transfers)
+        }
+
+        return walletStateMap.map { WalletState(it.key, it.value) }.toSet()
+    }
+
+    private fun calculateBalances(walletStateMap: MutableMap<String, BigDecimal>, transfers: List<TransferEventData>) {
+        transfers.forEach {
+            val toAmount = walletStateMap.getOrDefault(it.toAddress, BigDecimal.ZERO)
+
+            walletStateMap[it.toAddress] = toAmount.plus(it.amount)
+
```

Comment:
Need to define const list with ignored addresses and check for contains here

Reply:
ok, i will extract to constant variable, but i don't think more addresses will be added